packages/block-directory/src/store/test/actions.js

```diff
@@ -25,288 +64,218 @@ describe( 'actions', () => {
-			expect( generator.next() ).toEqual( {
-				value: false,
-				done: true,
-			} );
+		it( "should set an error if the plugin can't install", async () => {
+			const registry = createRegistryWithStores();
+
+			// mock the api-fetch and load-assets modules
+			apiFetch.mockImplementation( async ( { path } ) => {
+				switch ( path ) {
+					case 'wp/v2/plugins':
+						throw {
+							code: 'plugins_api_failed',
+							message: 'Plugin not found.',
+							data: null,
+						};
+					default:
+						throw new Error( `unexpected API endpoint: ${ path }` );
+				}
+			} );
+
+			loadAssets.mockImplementation( loadAssetsMock( registry ) );
+
+			// install the block
+			await registry
+				.dispatch( blockDirectoryStore )
+				.installBlockType( block );
+
+			// check that blocks store doesn't contain the new block
+			const registeredBlock = registry
+				.select( blocksStore )
+				.getBlockType( 'block/block' );
+			expect( registeredBlock ).toBeUndefined();
+
+			// check that error notice was displayed
+			const notices = registry.select( noticesStore ).getNotices();
+			expect( notices ).toMatchObject( [
+				{ content: 'Plugin not found.' },
+			] );
 		} );
 	} );
 
 	describe( 'uninstallBlockType', () => {
-		const block = {
-			...item,
-			links: {
-				...item.links,
-				self: [
-					{
-						href: pluginEndpoint,
-					},
-				],
-			},
-		};
-
-		it( 'should uninstall a block successfully', () => {
-			const generator = uninstallBlockType( block );
-
-			// First the deactivation step
-			expect( generator.next().value ).toMatchObject( {
-				type: 'API_FETCH',
-				request: {
-					url: pluginEndpoint,
-					method: 'PUT',
-				},
-			} );
-
-			// Then the deletion step
-			expect( generator.next().value ).toMatchObject( {
-				type: 'API_FETCH',
-				request: {
-					url: pluginEndpoint,
-					method: 'DELETE',
-				},
-			} );
-
-			expect( generator.next().value ).toEqual( {
-				type: 'REMOVE_INSTALLED_BLOCK_TYPE',
-				item: block,
-			} );
-
-			expect( generator.next() ).toEqual( {
-				value: undefined,
-				done: true,
-			} );
+		const installedBlock = blockWithLinks( block, {
+			self: [ { href: pluginEndpoint } ],
 		} );
 
-		it( "should set a global notice if the plugin can't be deleted", () => {
-			const generator = uninstallBlockType( block );
-
-			expect( generator.next().value ).toMatchObject( {
-				type: 'API_FETCH',
-				request: {
-					url: pluginEndpoint,
-					method: 'PUT',
-				},
-			} );
+		it( 'should uninstall a block successfully', async () => {
+			const registry = createRegistryWithStores();
+
+			apiFetch.mockImplementation( async ( { url, method } ) => {
+				switch ( url ) {
+					case pluginEndpoint:
+						switch ( method ) {
+							case 'PUT':
+							case 'DELETE':
+								return;
+							default:
+								throw new Error(
+									`unexpected API endpoint method: ${ method }`
+								);
+						}
+					default:
+						throw new Error( `unexpected API endpoint: ${ url }` );
+				}
```

Comment:
That's a lot of nesting :-) How about this?

```suggestion
				if ( url !== pluginEndpoint) {
					throw new Error( `unexpected API endpoint: ${ url }` );
				}
				
				if ( ! [ 'PUT', 'DELETE' ].includes( method ) ) {
					throw new Error( `unexpected API endpoint method: ${ method }` );
				}
```

Reply:
I was considering exactly that, but then I decided that I like the `switch` statements better, even though they are more nested and verbose. The reason is that the `switch` structure is much closer to how a real HTTP server would look like, with `case` statements for each supported route, then for each supported method...

Reply:
You have a point. I still prefer the ifs, but that's ultimately a matter of taste. Let's keep that `switch` then.